aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -294,7 +298,34 @@ inline Return Dispatcher::callWithDispatchKey(const OperatorHandle& op, Dispatch
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
   const auto& dispatchTable = op.operatorIterator_->op.dispatch_table();
   const KernelFunction& kernel = dispatch_(dispatchTable, dispatchKey);
-  return kernel.template call<Return, Args...>(op, std::forward<Args>(args)...);
+
+  if (dispatchKey != DispatchKey::BackendSelect) {
```

Comment:
This special case looks really suspicious. Why do you have to special case BackendSelect?

Reply:
I see in stack traces that for ops we call callWithDispatchKey first with BackendSelect
and then callWithDispatchKey with an actual dispatchKey

Reply:
Yes, this is what BackendSelect does: when you do a factory function, you hit BackendSelect first, and then BackendSelect figures out what the correct Backend is and then redispatches you there.

It seems clear you don't want to profile twice in this case. But there should be a very simple fix for this: make *BackendSelect* redispatch bypass record function. Then you can avoid this special case entirely.

Reply:
what do you mean by "make BackendSelect redispatch bypass record function"? Right now we don't use RecordFunction in `redispatch`

Reply:
Sorry, I wasn't using redispatch in the technical sense here. Here's what BackendSelect generates right now:

```
Tensor _cudnn_init_dropout_state(double dropout, bool train, int64_t dropout_seed, const TensorOptions & options) {
  static auto op = c10::Dispatcher::singleton()
    .findSchemaOrThrow("aten::_cudnn_init_dropout_state", "")
    .typed<Tensor (double, bool, int64_t, const TensorOptions &)>();
  DispatchKey _dk = options.computeDispatchKey();
  return op.callWithDispatchKey(_dk, dropout, train, dropout_seed, options);
}
```

You should replace callWithDispatchKey with a call that bypasses RecordFunction.

Reply:
I assume, you mean passing a flag into callWithDispatchKey? e.g. extra boolean whether to use record function

Reply:
Mm, the exact API is up to you. It might be better to make it an optional template parameter so that it is obvious that it is statically known (and so the condition will be optimized away).

Reply:
@ezyang it feels kind of inverted to make avoiding double counting the responsibility of kernels, doesn't it? 

If we can say something simple like: we only want to record invocations that run "real" kernels (dispatch key = CPU, GPU, ...), and not others, then wouldn't it be better to have some predicate `shouldRecord(dk)` on dispatch keys that we can call from the e.g. site of @ilia-cher 's original if test? Centralizing the knowledge in the dispatcher would
* avoid the entanglement where "nonterminal" kernels (e.g. BackendSelect, but maybe others as well - now or later) need to know about RecordFunction and double counting
* avoid muddy reentrancy situations e.g. where suppressing RecordFunction in BackendSelect winds up undercounting reentrant invocations issued from whatever kernel is selected. (Because in the current solution, we're counting the BackendSelect invocation, not the subsequent kernel invocation.)

Reply:
There are a few reasons I don't like it:

1. Now we have to do an extra computation and test to see if we do record function. If kernels decide whether or not to suppress record function or not, you can eliminate the conditional. This is my most important reason.
2. Testing on dispatch key is incompatible with fallthrough optimization. The semantics of fallthrough are that it should be equivalent to having invoked on the dispatch key, and then the kernel redispatched to the next kernel. With the fallthrough optimization, there is never actually an invocation of callWithDispatchKey with the dispatch key that was supposed to fall through, which means if both dispatch keys were intended to RecordFunction, then you will lose one of the observer calls. If you say that dispatcher calls can decide if they record function or not, then you can simply define the reference implementation of fallthrough to bypass profiling.

I also feel that RecordFunction suppression morally belongs at these call site. A call site that is redispatching knows if it did "a lot of" work or not very much work. When a call site has done very little work, it should skip RecordFunction and let the redispatch attribute it costs to the enclosing cost center. If initial calls into the dispatcher always do RecordFunction, there will always be a reasonable cost center to attribute the performance too.

Reply:
re. 1 - isn't it a trivial check (uint8_t == check?)

Reply:
a qq. - BackendSelect is not supposed to happen in callBoxed?

Reply:
@ezyang @ilia-cher sorry for the late reply -
1. true, it's an extra conditional, but only when RecordFunction is on. (I just realized @ilia-cher 's code here doesn't do it this way, but I think it could - i.e., line 336 here could be moved to after line 344.) So we'd be making profiling incrementally faster by avoiding the conditional, not all dispatch.
2. hmm but isn't it incompatible with fallthrough to host profiling in callWithDispatchKey, irrespective of filtering? You'll miss the obvservations regardless. If I follow your description, I think you're just saying that if the convention is for kernels to decide, then you avoid this by... having kernels instead of fallthrough? I'm probably missing your point here.

Re morality, I agree that call-site-decides is one way to bucket a multi-level dispatch chain together. I think for things like BackendSelect, where as a class of kernels they're going to be predictably lightweight, you have the information you need to make a centralized decision - but maybe it's more fraught to assume the same for keys with more diverse kernels.

Anyway my most important reason :) is wanting to keep arcane dispatcher knowledge from leaking out into kernels. Like, will each new set of kernels for utility dispatch keys (e.g., Meta) need to remember to turn profiling off? Will observers ever be used for new things for which would kernels would need to make new decisions? But I guess we can always go back and revisit if any of that starts to become problematic.

Reply:
@bhosmer do you mean that it's better in your opinion to always create `at::RecordFunction guard(at::RecordScope::FUNCTION);`, then check `if (C10_UNLIKELY(guard.active))` and then for `useRecordFunction && op.operatorIterator_->op.isObserved()` ? wouldn't it be better to just avoid this path completely if useRecordFunction is turned off or the op is not observable at all?

Reply:
also I'm more inclined to agree with @bhosmer as it does seem more to expose the detail of dispatcher to the interface,  will wait for @ezyang to address

Reply:
> @bhosmer do you mean that it's better in your opinion to always create `at::RecordFunction guard(at::RecordScope::FUNCTION);`, then check `if (C10_UNLIKELY(guard.active))` and then for `useRecordFunction && op.operatorIterator_->op.isObserved()` ? wouldn't it be better to just avoid this path completely if useRecordFunction is turned off or the op is not observable at all?

Well, in my mind useRecordFunction is the thing we're trying to avoid, right? (Aside - if we do go with useRecordFunction, I'd make it a template param per @ezyang's suggestion. The passed arg is always constant AFAICT, and the specialization should give us a nice perf bump.)

Anyway though, what I was noting was that the logic went from
```
at::RecordFunction guard(at::RecordScope::FUNCTION);
if (C10_UNLIKELY(guard.active)) {
  do profiling stuff 
}
```
to
```
if (dispatchKey != BackendSelect) { 
  at::RecordFunction guard(at::RecordScope::FUNCTION);
  if (C10_UNLIKELY(guard.active)) {
    do profiling stuff 
  }
}
```
which adds the comparison unconditionally to a very hot hot path, which is definitely something we'd want to avoid if we can. But if instead we move the new test inside the record function block (logically equivalent I think), then we're  only adding overhead when we're profiling:
```
at::RecordFunction guard(at::RecordScope::FUNCTION);
if (C10_UNLIKELY(guard.active)) {
  if (dispatchKey != BackendSelect) { // or e.g. shouldRecord(dispatchKey) or whatever
    do profiling stuff 
  }
}
```